### DISCUSSION TITLE
How to register beans by annotation members programatically

### DISCUSSION THREAD

**Discussion by subash89** [micronaut-projects/micronaut-core#8159] [General] [answered]
I have a class and need to inject multiple objects of it based on a custom annotation, say @MyJob. I can I achieve something like following ?  Appreciate any help around this. Most AOP examples I see in docs is with method interceptor. Not like a field level as follows.

```
@Service
public class ServiceX {

  @MyJob("key1")
  MYJobImpl obj1;
  
  @MyJob("key2")
  MYJobImpl obj2;
}

public class MyJobImpl{
  String name;
  public MyJobImpl(String name){
   this.name=name
 }
}
```

**Comment by shingjo** [comment, marked as answer]
Something like [5.2 Introduction Advice](https://docs.micronaut.io/latest/guide/#introductionAdvice)?

**Reply by subash89** [reply]
Thank you very much for the prompt response. 
I looked at that as well, but the example is with a method interceptor. But how can I inject those into a bean as fields as in above example.

**Reply by shingjo** [reply]
The example is showing that:
```java
public interface StubExample {

    @Stub("10")
    int getNumber();

    LocalDateTime getDate();
}
```

**Reply by subash89** [reply]
It's an example of interface where this Stub AOP will provide implementation for getNumber() method in an implementation for StubExample interface.  Not a field. 
Basically I am trying to inject different objects based on an annotation and its members.

**Reply by shingjo** [reply]
You can use a Factory to create your beans.

```java
public class MyJobImpl{
    String name;
    public MyJobImpl(String name){
        this.name=name;
    }
}

@Factory
public class MyJobFactory {

    @Prototype
    public MyJobImpl createJob(InjectionPoint<?> injectionPoint) {
        Optional<String> type = injectionPoint.getAnnotationMetadata().stringValue(MyJob.class);

        if(type.isPresent()) {
            return new MyJobImpl(type.get());
        } else {
            return null;
        }
    }
}

public class ServiceX {
    @Inject
    @MyJob("key1")
    MyJobImpl obj1;

    @Inject
    @MyJob("key2")
    MyJobImpl obj2;
}
```

**Reply by subash89** [reply]
This looks promising. Let me try this.

**Reply by subash89** [reply]
@shingjo That worked perfectly well. Thanks a lot.

**Reply by subash89** [reply]
@shingjo  
One quick last question on this for a different use case. Any chance we can use same Factory way to @Named bind different objects instead of custom binding.  
Currently how I do is manually create the objects at startup and `ctx.registerSingleton(MyJobImpl.class, myjob1, Qualifiers.byName("myjob1"));`


I tried to follow same model you suggested above, assuming @Named vs @MyJob are pretty much same, some annotation, so it should behave same. But when @Named is present, the Factory method(wtih injectionPoint)  not getting invoked.
